turbo/stages/headerdownload/header_algos.go

```diff
@@ -682,6 +718,75 @@ func (hd *HeaderDownload) InsertHeaders(hf func(header *types.Header, headerRaw
 	return hd.highestInDb >= hd.preverifiedHeight && hd.topSeenHeight > 0 && hd.highestInDb >= hd.topSeenHeight, nil
 }
 
+func (hd *HeaderDownload) InsertHeadersBackwards(tx kv.RwTx, canonicalHashCollector *etl.Collector, headerCollector *etl.Collector, logPrefix string, logChannel <-chan time.Time) (bool, error) {
+	if len(hd.insertList) == 0 {
+		return false, nil
+	}
+	hd.lock.Lock()
+	defer hd.lock.Unlock()
+	sort.Slice(hd.insertList, func(i, j int) bool {
+		return hd.insertList[i].header.Number.Uint64() > hd.insertList[j].header.Number.Uint64()
+	})
+	for len(hd.insertList) > 0 {
+		header := hd.insertList[0].header
+		if header.Number.Uint64() >= hd.lastProcessedPayload {
+			hd.insertList = hd.insertList[1:]
+			continue
+		}
+		// If we miss some headers or an header results to be invalid, we ask for them again
+		if header.Number.Uint64() != hd.lastProcessedPayload-1 {
+			break
+		}
+		if header.Hash() != hd.expectedHash {
+			hd.insertList = hd.insertList[1:]
+			break
+		}
+		// Write the encoded header
+		data, err := rlp.EncodeToBytes(header)
+		if err != nil {
+			log.Crit("Failed to RLP encode header", "err", err)
+		}
+		if err := headerCollector.Collect(dbutils.HeaderKey(header.Number.Uint64(), header.Hash()), data); err != nil {
+			return false, err
+		}
+		if err := canonicalHashCollector.Collect(dbutils.EncodeBlockNumber(header.Number.Uint64()), header.Hash().Bytes()); err != nil {
+			return false, err
+		}
+		hd.lastProcessedPayload--
+
+		canonical, err := rawdb.ReadCanonicalHash(tx, hd.lastProcessedPayload-1)
+		if err != nil {
+			return false, err
+		}
+		if canonical == header.ParentHash {
+			return true, nil
+		}
+		hd.expectedHash = header.ParentHash
+		hd.insertList = hd.insertList[1:]
+	}
+	hd.missingBlockNumber = hd.lastProcessedPayload - 1
+	return false, nil
+}
+
+func (hd *HeaderDownload) SetExpectedHash(hash common.Hash) {
+	hd.lock.Lock()
+	defer hd.lock.Unlock()
+	hd.expectedHash = hash
+}
+
+func (hd *HeaderDownload) ProcessSegmentPOS(segment ChainSegment) {
+	hd.lock.Lock()
+	defer hd.lock.Unlock()
+	log.Trace("Appending...", "from", segment[0].Number, "to", segment[len(segment)-1].Number, "len", len(segment))
+	for _, segmentFragment := range segment {
+		// If we found the block number we were missing, we can just dismiss it
+		if segmentFragment.Header.Number.Uint64() == hd.missingBlockNumber {
+			hd.missingBlockNumber = 0
+		}
+		hd.insertList = append(hd.insertList, &Link{header: segmentFragment.Header})
```

Comment:
you can bypass `insertList` and just do insertion into the collectors right here. Basically, move the content of `InsertHeadersBackwards` into this one. Since headers don't go directly into DB, but into collectors, it should be pretty quick, and you don't need to allocate more objects for links

Reply:
i also want to check if the headers match with their parenthashes to check their validity, so i need to process them in order, not all requests are received in the order we sent them, i still will need to keep them somewhere.

Reply:
that is why, i first wait for the previous request to go in before moving on

Reply:
Ok, understood. One small improvement you can make is to add `headerRaw` into the link you are creating, so you don't have to do RLP encoding again in the `InsertHeadersBackwards`. I introduced this after you've first created this PR